### PR TITLE
Add wine version and wine colors to overlay

### PR DESCRIPTION
**Pull request by gort818**
Allows the ability to show current version running, and to change the color. Not functioning with proton versions yet.

### CODE REVIEW

**Comment by rawfoxDE** [comment]
Good job man !

**Comment by jackun** [comment]
Does using `get_exe_path()` and getting `wine` path from `wine[64]-preloader`  aka /proc/self/exe work?

**Comment by gort818** [comment]
I will give that a shot.
Sure no problem.

File: src/overlay.cpp [hunk 1]
```diff
@@ -1301,6 +1325,15 @@ void render_imgui(swapchain_stats& data, struct overlay_params& params, ImVec2&
          ImGui::PopFont();
       }
 
+       if (params.enabled[OVERLAY_PARAM_ENABLED_wine]){
+           ImGui::TextColored(ImGui::ColorConvertU32ToFloat4(params.wine_color), "%s","WINE");
+           ImGui::PushFont(data.font1);
+           //ImGui::Dummy(ImVec2(0, 8.0f));
+           auto wine_color = ImGui::ColorConvertU32ToFloat4(params.wine_color);
```
Comment (flightlessmango) [inline src/overlay.cpp:1332]: No need to convert the `wine_color` twice

**Comment by gort818** [comment]
Well I believe I screwed up that rebase. Should I just work off the develop branch then submit another PR to merge into develop?

**Comment by jackun** [comment]
:D Just `git rebase -i 5ce6bcb79cb0dd1de41c79fcfa8a410d0dde645e` and drop the other unrelated commits, probably.

**Comment by gort818** [comment]
Well I just made it worse :( so sorry

**Comment by gort818** [comment]
Well that is a little better, not sure how that readme change ended up there..oops

**Comment by jackun** [comment]
"Merge branch 'develop' into wine2" should be probably squashed into "Add wine and wine colors to overlay"

File: src/overlay.cpp [hunk 2]
```diff
@@ -1341,6 +1365,13 @@ void render_imgui(swapchain_stats& data, struct overlay_params& params, ImVec2&
          ImGui::PopFont();
       }
 
+       if (params.enabled[OVERLAY_PARAM_ENABLED_wine]){
+           ImGui::TextColored(ImGui::ColorConvertU32ToFloat4(params.wine_color), "%s","WINE");
+           ImGui::PushFont(data.font1);
+           ImGui::TextColored(wine_color, "%s", wineVersion.c_str());
```
Comment (flightlessmango) [inline src/overlay.cpp:1371]: `wine_color` doesn't exist now, the first conversion is not stored

**Comment by gort818** [comment]
Ok, I think I got it squared away. Thanks for the help!

**Comment by flightlessmango** [comment]
It would be good if this doesn't show when it's not wine

**Comment by gort818** [comment]
At the moment it will display "N/A" if wine is enabled in the config at not runnig, you want it to not display at all if wine is not being used?

**Comment by jackun** [comment]
It is probably better to use `get_exe_name()` and derive `wine`'s path from it. There could be bajillion other wineservers running.

**Comment by gort818** [comment]
Yeah was testing that out this past weekend, I will get something together.

**Comment by flightlessmango** [comment]
> At the moment it will display "N/A" if wine is enabled in the config at not runnig, you want it to not display at all if wine is not being used?

Yes, I think that would be better

**Comment by gort818** [comment]
Awesome, thanks for clarifying! I will get on it.

**Comment by gort818** [comment]
Working on getting Proton version to display as well, should have it ready shortly.

**Comment by gort818** [comment]
When the wine option is enable here is how it looks.

MangoHud showing wine version running Star Citizen
wine-mango
MangoHud not showing wine version running vkcube
mango-native
MangoHud showing Proton version running Doom Eternal 
mango-proton

File: src/overlay.cpp [hunk 3]
```diff
@@ -689,6 +689,52 @@ void init_system_info(){
       trim(gpu);
       driver = exec("glxinfo | grep 'OpenGL version' | sed 's/^.*: //' | cut -d' ' --output-delimiter=$'\n' -f1- | grep -v '(' | grep -v ')' | tr '\n' ' ' | cut -c 1-");
       trim(driver);
+
+// Get WINE version
+
+      wineProcess = get_exe_path();
+      trim(wineProcess);
```
Comment (jackun) [inline src/overlay.cpp:696]: Shouldn't be needed probably. Any whitespace is actually in the file's name then.

**Comment by jackun** [comment]
Too much `sed` for my liking :D Maybe something like this: 
[wine-version.diff.txt](https://github.com/flightlessmango/MangoHud/files/5074579/wine-version.diff.txt)

File: src/overlay.cpp [hunk 4]
```diff
@@ -694,24 +694,40 @@ void init_system_info(){
...
+      if (preloader == "wine-preloader" || preloader == "wine64-preloader") {
+         // Check if using Proton
+         if (wineProcess.find(searchProton) != std::string::npos) {
+            stringstream protonPath;
+            protonPath << "echo " << wineProcess <<" | rev | cut -d '/' -f3- | rev | sed 's/.*/\"&\"/'";
+            string protonVersion;
+            protonVersion = exec(protonPath.str());
+            trim(protonVersion);
+            stringstream command;
+            command <<"< " <<protonVersion + "/version" <<" cut -d' ' -f2";
```
Comment (flightlessmango) [inline src/overlay.cpp:709]: Thinking something like this to make it prettier and conform more to how it's displayed in Steam
```
command <<"< " <<protonVersion + "/version" <<" cut -d' ' -f2 | sed -e 's|proton-||' ";
wineVersion = exec(command.str());
wineVersion = "Proton " + wineVersion;
```

**Comment by gort818** [comment]
I might have gone a little overboard with sed:). Awesome I will take a look.

File: src/overlay.cpp (hunk 4 above)
Comment (gort818) [reply src/overlay.cpp:709]: Yes that is much nicer.. also was thinking what if we replace WINE with PROTON and then  display the proton version underneath?
`` ImGui::TextColored(wine_color, "%s", "PROTON");``

Comment (flightlessmango) [reply src/overlay.cpp:709]: I'm not sure the WINE/PROTON text is required, the version should always contain proton or wine in it already

**Comment by gort818** [comment]
MangoHud full with custom proton
mangonew
MangoHud full with default proton
mangonew1

**Comment by flightlessmango** [comment]
Everything looks a-ok to me :ok_hand:

**Comment by gort818** [comment]
Great to hear!

**Comment by jackun** [comment]
Should be in https://github.com/flightlessmango/MangoHud/commit/632f1e7000a68602ce3410ab987a130953b5b7d7

**Pull request closed without merging**